src/github.com/matrix-org/go-neb/services/google/google.go

```diff
@@ -0,0 +1,221 @@
+// Package google implements a Service which adds !commands for Google custom search engine.
+// Initially this package just supports image search but could be expanded to provide other functionality provided by the Google custom search engine API - https://developers.google.com/custom-search/json-api/v1/overview
+package google
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/go-neb/types"
+	"github.com/matrix-org/gomatrix"
+)
+
+// ServiceType of the Google service
+const ServiceType = "google"
+
+var httpClient = &http.Client{}
+
+// Unsused -- leaving this in place for the time being to show structure of the request
+// type googleQuery struct {
+// 	// Query search text
+// 	Query string `json:"q"`
+// 	// Number of search results
+// 	Num int `json:"num"`
+// 	// Search result offset
+// 	Start int `json:"start"`
+// 	// Size of images to serch for (usually set to "medium")
+// 	ImgSize string `json:"imgSize"`
+// 	// Type of search - Currently always set to "image"
+// 	SearchType string `json:"searchType"`
+// 	// Type of image file to retur64 `json:"totalResults"`
+// 	FileType string `json:"fileType"`
+// 	// API key
+// 	Key string `json:"key"`
+// 	// Custom serch engine ID
+// 	Cx string `json:"cx"`
+// }
+
+type googleSearchResults struct {
+	SearchInformation struct {
+		TotalResults int64 `json:"totalResults,string"`
+	} `json:"searchInformation"`
+	Items []googleSearchResult `json:"items"`
+}
+
+type googleSearchResult struct {
+	Title       string      `json:"title"`
+	HTMLTitle   string      `json:"htmlTitle"`
+	Link        string      `json:"link"`
+	DisplayLink string      `json:"displayLink"`
+	Snippet     string      `json:"snippet"`
+	HTMLSnippet string      `json:"htmlSnippet"`
+	Mime        string      `json:"mime"`
+	FileFormat  string      `json:"fileFormat"`
+	Image       googleImage `json:"image"`
+}
+
+type googleImage struct {
+	ContextLink     string  `json:"contextLink"`
+	Height          float64 `json:"height"`
+	Width           float64 `json:"width"`
+	ByteSize        int64   `json:"byteSize"`
+	ThumbnailLink   string  `json:"thumbnailLink"`
+	ThumbnailHeight float64 `json:"thumbnailHeight"`
+	ThumbnailWidth  float64 `json:"thumbnailWidth"`
+}
+
+// Service contains the Config fields for the Google service.
+// TODO - move the google custom search engine ID in here!
+//
+// Example request:
+//   {
+//       "api_key": "AIzaSyA4FD39m9pN-hiYf2NRU9x9cOv5tekRDvM"
+//   }
+type Service struct {
+	types.DefaultService
+	// The Google API key to use when making HTTP requests to Google.
+	APIKey string `json:"api_key"`
+}
+
+// Commands supported:
+//    !google some search query without quotes
+// Responds with a suitable image into the same room as the command.
+func (s *Service) Commands(client *gomatrix.Client) []types.Command {
+	return []types.Command{
+		types.Command{
+			Path: []string{"google"},
```

Comment:
@ara4n, but the underlying api is just generic Google search. Wouldn't it make more sense just to extend this bot, rather than creating a new one that's 98% the same?